llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,141 @@
+// function, and moves the string literal to constant address
+// space. This a temporary solution for printf's support of generic
+// address space literals; the pass should be dropped once SYCL device
+// backends learn to handle the generic address-spaced argument properly.
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/MutatePrintfAddrspace.h"
+
+#include "llvm/Analysis/ValueTracking.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/InitializePasses.h"
+
+using namespace llvm;
+
+namespace {
+constexpr unsigned ConstantAddrspaceID = 2;
+
+// Wrapper for the pass to make it working with the old pass manager
+class SYCLMutatePrintfAddrspaceLegacyPass : public ModulePass {
+public:
+  static char ID;
+  SYCLMutatePrintfAddrspaceLegacyPass() : ModulePass(ID) {
+    initializeSYCLMutatePrintfAddrspaceLegacyPassPass(
+        *PassRegistry::getPassRegistry());
+  }
+
+  // run the SYCLMutatePrintfAddrspace pass on the specified module
+  bool runOnModule(Module &M) override {
+    ModuleAnalysisManager MAM;
+    auto PA = Impl.run(M, MAM);
+    return !PA.areAllPreserved();
+  }
+
+private:
+  SYCLMutatePrintfAddrspacePass Impl;
+};
+} // namespace
+
+char SYCLMutatePrintfAddrspaceLegacyPass::ID = 0;
+INITIALIZE_PASS(SYCLMutatePrintfAddrspaceLegacyPass,
+                "SYCLMutatePrintfAddrspace",
+                "Move SYCL printf literal arguments to constant address space",
+                false, false)
+
+// Public interface to the SYCLMutatePrintfAddrspacePass.
+ModulePass *llvm::createSYCLMutatePrintfAddrspaceLegacyPass() {
+  return new SYCLMutatePrintfAddrspaceLegacyPass();
+}
+
+static Constant *getCASLiteral(Module &M, IRBuilder<> &Builder, CallInst *CI) {
+  auto *Literal = dyn_cast<Constant>(CI->getArgOperand(0));
+  assert(Literal);
+  Type *LiteralType = Literal->getType();
+  // Generate/find a correct literal
+  auto *CASLiteralType = PointerType::get(LiteralType->getPointerElementType(),
+                                          ConstantAddrspaceID);
+  StringRef CASLiteralName(Literal->getName().str() + "._AS2");
+  return M.getOrInsertGlobal(CASLiteralName, CASLiteralType, [&] {
+    StringRef LiteralValue;
+    getConstantStringInfo(Literal, LiteralValue);
+    GlobalVariable *GV = Builder.CreateGlobalString(
+        LiteralValue, CASLiteralName, ConstantAddrspaceID, &M);
+    GV->setLinkage(GlobalValue::LinkageTypes::InternalLinkage);
+    GV->setUnnamedAddr(GlobalValue::UnnamedAddr::None);
+    return GV;
+  });
+}
+
+static FunctionCallee getCASPrintfFunction(Module &M,
+                                           const Function &GenericASPrintfFunc,
+                                           Type *CASLiteralTy) {
+  auto *CASPrintfFuncTy = FunctionType::get(GenericASPrintfFunc.getReturnType(),
+                                            CASLiteralTy, /*isVarArg=*/true);
```

Comment:
It does end up as variadic in the IR, though.

Reply:
UPD: I should actually rework this to consider the case of `__SYCL_USE_NON_VARIADIC_SPIRV_OCL_PRINTF__`

Reply:
Done as per our offline discussion. The single variadic function does the trick both for non-variadic and variadic versions coming from the headers. Also, I believe hardcoding the generated variadic function name is OK since we're quite confident this is merely a temporary solution & the user-visible API for constant AS is unlikely to change anytime soon.